File: backend/app/schemas/portal.py
from pydantic import BaseModel
from typing import Any, Optional

class SelectOption(BaseModel):
    label: str
    value: str

class TableColumn(BaseModel):
    key: str
    title: str

class FormField(BaseModel):
    name: str
    label: str
    type: str
    required: bool = False
    placeholder: Optional[str] = None
    options: list[SelectOption] = []
    section: Optional[str] = None
    columns: list[TableColumn] = []
    default_rows: list[dict[str, str]] = []
    visible_when_field: Optional[str] = None
    visible_when_value: Optional[str] = None

class FormTemplate(BaseModel):
    id: str
    title: str
    department_id: str
    section_id: str
    fields: list[FormField]

class Section(BaseModel):
    id: str
    title: str
    form_id: str

class Department(BaseModel):
    id: str
    title: str
    icon: Optional[str] = None
    sections: list[Section]

File: frontend/src/config/portal.ts
export type FieldType = 'text' | 'textarea' | 'select' | 'date' | 'file' | 'email' | 'table';

export type SelectOption = {
  label: string;
  value: string;
};

export type TableColumn = {
  key: string;
  title: string;
};

export type FormField = {
  name: string;
  label: string;
  type: FieldType;
  required?: boolean;
  placeholder?: string;
  options?: SelectOption[];
  section?: string;
  columns?: TableColumn[];
  default_rows?: Record<string, string>[];
  /** For date fields: literal YYYY/MM/DD or `'today'` */
  maxDate?: 'today' | string;
  minDate?: string;
};

export type FormTemplate = {
  id: string;
  title: string;
  department_id?: string;
  section_id?: string;
  fields: FormField[];
};

export type Section = {
  id: string;
  title: string;
  form_id: string;
};

export type Department = {
  id: string;
  title: string;
  icon?: string;
  sections: Section[];
};

export const API_BASE = "/api/v1";
